### PR TITLE
fix(runtime): keep exclusive interval boundaries exclusive

### PR DESCRIPTION
**Pull request by alexzautke**
## Description

The `Interval` selector normalized an exclusive date, dateTime or time boundary into a closed one by stepping it one unit of the boundary value's own precision inward. That is lossless only when the later comparison happens at that same precision. Timing phrases such as `6 months or less before day of start of X` build an interval with an exclusive high and compare with `in ... day`, so the pre-closed boundary (one millisecond before the start of the day) turned the exclusive same-day comparison into an inclusive one. `In` already implements the exclusive comparison the specification requires (CQL 1.5.3 Appendix B, section "In": "For open interval boundaries, exclusive comparison operators are used") but never saw an open boundary.

The selector now returns the interval with its boundary flags intact, and `Contains` shares `In`'s implementation (the spec states both in the same words, and calls the point overload of `Includes` a synonym of `contains`), so `in`, `contains` and `includes` agree at a coarser precision.

**Operators defined through Start/End.** With exclusive boundaries reaching the operators, everything the spec defines "using the semantics described in the Start and End operators" has to compare effective boundaries rather than raw endpoints; the selector's normalization used to provide that implicitly. Those operators now normalize their operands with `ToClosed` at entry: `Starts`, `Ends`, `Includes`/`Included In`, `Properly Includes`/`Properly Included In` (interval-interval and point-interval overloads), `Intersect`, `Point From`, and the ordering `Collapse` merges in. Unchanged, because they already normalized: `After`/`Before`, `Meets*`, `Overlaps*`, `SameAs`, `SameOr*`, `Except`, `Union`, `Expand`, `Width`, `Size`, and interval equality/equivalence/hashing in the comparer. `Start`/`End` keep stepping by one point unit.

**FHIR conversion.** The conversion of a CQL interval to a FHIR `Period` steps an exclusive date/time boundary inward by one unit of its precision, as the `Range` conversion already did, so emitted Periods do not change. Where that step is not representable (`CqlDate`/`CqlDateTime` at the end of their range return null; `CqlTime` wraps around midnight or yields negative components) the boundary is emitted unstepped rather than throwing or emitting an invalid value; omitting it would misstate a bounded interval as unbounded, and such an interval is degenerate anyway since its own `Start`/`End` is null for the same reason.

**CQL evaluation results change** for same-day boundary cases and for the operators listed above when an exclusive boundary is involved. Release-note fragment under `docs/releases/vnext/`. No generated code changes, so `GeneratorToolVersion` is untouched.

## Related issues

None filed. Found while tracing MADiE integration-runner failures for CMS646 (test cases `10cec7db`, `6e9a1974`, `ab48e0c0`: cystectomy on the day of staging, BCG on the day staging started).

## Testing

- `ExclusiveIntervalBoundaryPrecisionTests` (5 tests: selector preserves flags; `in ... day` excludes the boundary day for an exclusive high and an exclusive low; `in` without precision still compares at full precision; `Start`/`End` still step). Three fail without the runtime change.
- `EffectiveIntervalBoundaryOperatorTests` (13 tests): the exclusive-boundary case for each operator above, plus `contains` agreeing with `in` at day precision. Twelve fail without the operator changes; the thirteenth guards that `point from` still throws for an interval wider than one point.
- `FhirTypeConverterHostConversionTests`: Period conversion of exclusive and closed dateTime/date/time boundaries, plus five extrema cases; every emitted boundary validated with `FhirDateTime.IsValidValue`/`Date.IsValidValue`.
- `CoreTests`: 1405 passed, 0 failed, 8 skipped (net10.0).
- Integration runner, full QI-Core 2025 corpus in validate mode against `passed.jsonl`: the three CMS646 cases above start passing; no regressions.

## FirelyTeam Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] Mark the PR with the label **breaking change** when this PR introduces breaking changes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01AoPNvHfeXidST7F5ypteFd

### CODE REVIEW

File: Cql/Cql.Runtime/Operators/CqlOperators.IntervalOperators.cs [hunk 1]
```diff
@@ -78,9 +78,16 @@ internal partial class CqlOperators
...
+                // Boundary exclusivity is deliberately preserved rather than normalized away with
+                // ToClosed(): closing an exclusive date/time boundary shifts it by one unit of the
+                // boundary value's own precision, which is lossy for any operator that compares at a
+                // coarser precision. In: "For open interval boundaries, exclusive comparison operators
+                // are used. [...] If precision is specified and the point type is a date/time type,
+                // comparisons used in the operation are performed at the specified precision."
+                // (CQL 1.5.3 Errata 2, Appendix B - CQL Reference, 5.3 in). Pre-closing an exclusive
+                // high of @2026-06-30T08:00 to @2026-06-30T07:59:59.999 turns 'in ... day' from an
+                // exclusive same-day comparison into an inclusive one.
+                return new CqlInterval<CqlDateTime?>(low, high, lowClosed, highClosed);
```
[the remark this answers is not in the document]
Comment (alexzautke) [reply Cql/Cql.Runtime/Operators/CqlOperators.IntervalOperators.cs:90]: Confirmed, and the audit found the problem wider than the three named. Fixed in a869da3: every operator the spec defines "using the semantics described in the Start and End operators" now normalizes its operands with `ToClosed` at entry (via a `ToClosedForPointType<T>` dispatcher for the generic ones): `Starts`, `Ends`, `IntervalIncludesInterval`/`IncludedIn`, `IntervalProperlyIncludes`/`ProperlyIncludedIn` (interval-interval and all point-interval overloads), `Intersect` (read raw, `[a, b) intersect [b, c]` returned an empty interval instead of null), `PointFrom` (read raw, it threw for any open interval; spec example `point from Interval[4, 5) // 4`), and the sort key `Collapse` merges in. `Contains` now shares `In`'s implementation, since the spec states both in the same words and `Includes` calls the point overload a synonym of `contains`; it had kept the very defect this PR removes from `In`. `In` itself stays un-closed. Unchanged because they already normalize: `After`/`Before`, `Meets*`, `Overlaps*`, `SameAs`, `SameOr*`, `Except`, `Union`, `Expand`, `Width`, `Size`, and the interval comparer. `EffectiveIntervalBoundaryOperatorTests` pins the exclusive-boundary case for each changed operator; 12 of its 13 tests fail on the previous head.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Cql/Cql.Firely/FhirTypeConverter.cs [hunk 2]
```diff
@@ -362,6 +362,10 @@ internal static TypeConverter ConvertCqlPrimitivesToFhir(
             converter.AddConversion((CqlInterval<long?> interval) => interval is null
                 ? null
                 : NumericIntervalToRange(interval.low, interval.high, interval.lowClosed, interval.highClosed, 1m));
+            // A FHIR Period has inclusive boundaries, so an exclusive CQL boundary is stepped inward by one
+            // unit of its own precision, exactly like NumericIntervalToRange does for a Range. The CQL
+            // interval itself keeps its exclusivity (the Interval selector no longer normalizes it away),
+            // because operators comparing at a coarser precision need the original boundary.
```
[the remark this answers is not in the document]
Comment (alexzautke) [reply Cql/Cql.Firely/FhirTypeConverter.cs]: Confirmed empirically: `CqlDate.MaxValue.Successor()` and the `CqlDateTime` equivalent come back null through the `!` (so the old code dereferenced null), `CqlTime.MaxValue.Successor()` wraps to `00:00:00.000`, and `CqlTime.MinValue.Predecessor()` yields `00:00:00.-001`. Fixed in a869da3 with a `PeriodBoundary` helper per point type: an exclusive boundary whose step is not representable is emitted unstepped. Omitting it would turn a bounded interval into an unbounded `Period`, which misstates the value far more than one unit of its own precision, and such an interval is degenerate anyway (its own `Start`/`End` is null for the same reason, per the spec's Successor/Predecessor overflow rule). For `CqlTime` the step is accepted only when it stays a non-negative time of day that moved in the expected direction, which catches both the midnight wrap and the negative components. Five new extrema tests, all validated with `IsValidValue`; all five fail on the previous head.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: Cql/CoreTests/FhirTypeConverterHostConversionTests.cs [hunk 3]
```diff
@@ -24,6 +24,51 @@ public class FhirTypeConverterHostConversionTests
...
+            // before the high.
+            var low = new CqlDateTime(2026, 1, 1, 0, 0, 0, 0, 0, 0);
+            var high = new CqlDateTime(2026, 7, 1, 0, 0, 0, 0, 0, 0);
+            var interval = new CqlInterval<CqlDateTime>(low, high, lowClosed: false, highClosed: false);
+
+            var converted = FhirTypeConverter.Convert<Period>(interval);
+
+            Assert.IsNotNull(converted);
+            Assert.AreEqual("2026-01-01T00:00:00.001Z", converted.Start);
+            Assert.AreEqual("2026-06-30T23:59:59.999Z", converted.End);
```
[the remark this answers is not in the document]
Comment (alexzautke) [reply Cql/CoreTests/FhirTypeConverterHostConversionTests.cs:41]: Done in a869da3: the three tests, and every Period test added since, validate both boundaries through `FhirDateTime.IsValidValue` or `Date.IsValidValue` (matching the primitive each path emits) via two small helpers.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by ewoutkramer — APPROVED** [review]
